File: lib/wType/object.cpp
#include "object.h"

#include "string.h"
#include "vocabulary.h"
#include "uint64.h"
#include "address.h"
#include "boolean.h"
#include "event.h"
#include "vector.h"
#include "blob.h"
#include <arpa/inet.h>
//#include <stdint.h>

W::Object::Object()
{

}

W::Object::~Object()
{

}

W::Object* W::Object::fromByteArray(ByteArray& in)
{
    uint32_t type = in.pop8();
    
    Object *answer;
    
    switch (type)
    {
        case string:
            answer = new String();
            break;
            
        case vocabulary:
            answer = new Vocabulary();
            break;
            
        case uint64:
            answer = new Uint64();
            break;
            
        case address:
            answer = new Address();
            break;
            
        case boolean:
            answer = new Boolean();
            break;
            
        case event:
            answer = new Event();
            break;
            
        case vector:
            answer = new Vector();
            break;
            
        case blob:
            answer = new Blob();
            break;
    }
    
    answer->deserialize(in);
    return answer;
}

bool W::Object::sameType(const W::Object& other) const
{
    return getType() == other.getType();
}

bool W::Object::operator!=(const W::Object& other) const
{
    return !operator==(other);
}

W::Object::StdStr W::Object::getTypeName(W::Object::objectType type)
{
    switch (type) {
        case string:
            return "String";
        case vocabulary:
            return "Vocabulary";
            
        case uint64:
            return "Uint64";
            
        case address:
            return "Address";
            
        case boolean:
            return "Boolean";
            
        case event:
            return "Event";
            
        case vector:
            return "Vector";
            
        case blob:
            return "Blob";
    }
}

File: libjs/wType/object.js
"use strict";
var Class = require("../utils/class");
        
var Object = Class.inherit({
    "className": "Object",
    "constructor": function() {
        Class.fn.constructor.call(this);
    },
    "<": function(other) {
        throw new Error(this.className + " has no reimplemented method \"<\"");
    },
    ">": function(other) {
        throw new Error(this.className + " has no reimplemented method \">\"");
    },
    "==": function(other) {
        throw new Error(this.className + " has no reimplemented method \"==\"");
    },
    "clone": function() {
        throw new Error(this.className + " has no reimplemented method \"clone\"");
    },
    "getType": function() {
        var type = Object.objectType[this.className];
        
        if (type === undefined) {
            throw new Error("Undefined type of " + this.className);
        }
        
        return type;
    },
    "length": function() {
        throw new Error(this.className + " has no reimplemented method \"length\"");
    },
    "size": function() {
        throw new Error(this.className + " has no reimplemented method \"size\"");
    },
    "toString": function() {
        throw new Error(this.className + " has no reimplemented method \"toString\"");
    },
    "valueOf": function() {
        throw new Error(this.className + " has no reimplemented method \"valueOf\"");
    }
});

Object.objectType = {
    "String"    :   0,
    "Vocabulary":   1,
    "Uint64"    :   2,
    "Address"   :   3,
    "Boolean"   :   4,
    "Event"     :   5,
    "Vector"    :   6,
    "Blob"      :   7
};

Object.reverseObjectType = {
    0   :   "String",
    1   :   "Vocabulary",
    2   :   "Uint64",
    3   :   "Address",
    4   :   "Boolean",
    5   :   "Event",
    6   :   "Vector",
    7   :   "Blob"
}

Object.fromByteArray = function() {
    throw new Error("Initialization error. Object.fromByteArray is not implemented, it implements in factory.js");
}

module.exports = Object;
